### PR TITLE
expression: fix wrong type inferring for ceiling function. (#21920)

### PR DESCRIPTION
**Pull request by ti-srebot**
cherry-pick #21920 to release-4.0

---

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/16935 

Problem Summary: 

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Related changes

- Need to cherry-pick to the release branch 4.0,3.0

### Check List 

Tests 

- Unit test
- Integration test

Side effects

### Release note 

- Fix wrong type inferring for ceiling function.

### CODE REVIEW

File: ddl/db_partition_test.go [hunk 1]
```diff
@@ -334,7 +334,15 @@ func (s *testIntegrationSuite2) TestCreateTableWithHashPartition(c *C) {
                                        PARTITION p1 VALUES LESS THAN (200),
                                        PARTITION p2 VALUES LESS THAN MAXVALUE)`)
 	tk.MustGetErrCode("select * from t_sub partition (p0)", tmysql.ErrPartitionClauseOnNonpartitioned)
+<<<<<<< HEAD
```
Comment (lzmhhh123) [inline ddl/db_partition_test.go:337]: Please resolve conflicts.

**Review by qw4990 — COMMENTED** [review]
LGTM

**Comment by qw4990** [comment]
PTAL @lzmhhh123

**Comment by lzmhhh123** [comment]
LGTM

**Review by qw4990 — APPROVED** [review]

**Pull request merged**